### PR TITLE
Adding a function to calculate a derivative numerically

### PR DESCRIPTION
**Pull request by jjram20**
This commit includes:

- Creation of function derivativeNum in location src/function/numerical/derivativeNum.js
- Creation of test of this function in derivativeNumm in location test/unit-tests/function/numerical/derivativeNum.test.js
- Export of this function in file src/factoriesAny.js

### CODE REVIEW

File: src/function/numerical/derivativeNum.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+     * It gets the function whose derivative we want to calculate as a string
+     * and it is parsed and evaluated by the library.
+     *
+     * Syntax:
+     *
+     * math.derivativeNum(string, number)
+     *
+     * Example:
+     *
+     * math.derivativeNum('cos(x)', 5)  // returns
```
Comment (josdejong) [inline src/function/numerical/derivativeNum.js:20]: Can you add the actual result to the comment? (maybe pick these nice examples from your unit test)

**Comment by josdejong** [comment]
Thanks @jjram20 👍 

A few feedbacks:

- Can you also add embedded documention for the new function in the file [embeddedDocs.js](https://github.com/josdejong/mathjs/blob/develop/src/expression/embeddedDocs/embeddedDocs.js)?
- I think the function will be more powerful if we add a couple more types of inputs alongside the `derivativeNum(string, number)` that you've implemented:
    - `derivativeNum(string, number | BigNumber | Fraction)`  this is what you implemented right now, but we can accept other numeric types (BigNumber, Fraction)
    - `derivativeNum(Node, number | BigNumber | Fraction)` if you repeatedly want to evaluate the derivative, it is much faster to parse your expression only once and pass that to `derivativeNum` 
    - `derivativeNum(function, number | BigNumber | Fraction)` just allow any function as input, that will make the function very flexible

Does that make sense?

For reference: #2302

**Comment by josdejong** [comment]
@gwhitney thanks for your inputs. Two feedbacks:

1. I do not think the symbolic `derivative` is capable of calculating the derivative of every arbitrary function. In my experience you sometimes just need brute force number crunching for derivation and integration. 
2. Good point about the current simple implementation not being guaranteed to work reliably for each and every case, like around cases where a function is not continuous but goes to infinity. It may be a good enough start though, and it is very pragmatic to have I think. In next iterations we could improve on the algorithm, and maybe even built in warnings/errors when we detect that the derivative cannot reliable be calculated (like at `x=0` given `1/x`). It would be good to say a bit about the limitations of numerical derivation in the docs.

**Comment by gwhitney** [comment]
> 1. I do not think the symbolic `derivative` is capable of calculating the derivative of every arbitrary function. In my experience you sometimes just need brute force number crunching for derivation and integration.

Well, any function that can be coded in JavaScript can be differentiated by automatic differentiation: see https://github.com/ehaas/js-autodiff, https://github.com/iffsid/ad.js, https://tiarkrompf.github.io/notes/?/differentiable-programming-in-js/, and/or https://observablehq.com/@grjzwaan/building-autodiff-from-scratch. This goes doubly for mathjs functions, in which we already have the parse tree and can just traverse it for the forward pass and the reverse pass if we are doing that, and/or just do symbolic differentiation.

So the main time when you need to do numeric differentiation is when you don't have the ability to calculate values of the function, but can only observe them -- for example, if you are making physical measurements of something. But data/numerical analysis is at the moment not really in mathjs's scope, it seems much more centered on symbolic operation.

In any case, the consensus of current research seems to be that symbolic/automatic differentiation are pretty robust and numeric differentiation is pretty fragile and difficult, with the situation for the two approaches reversed for integration.

> 2. Good point about the current simple implementation not being guaranteed to work reliably for each and every case, like around cases where a function is not continuous but goes to infinity. It may be a good enough start though, and it is very pragmatic to have I think. In next iterations we could improve on the algorithm, and maybe even built in warnings/errors when we detect that the derivative cannot reliable be calculated (like at x=0 given 1/x). It would be good to say a bit about the limitations of numerical derivation in the docs.

From what I've read, the problem isn't just places where the function blows up. For functions that are changing slowly, the difference quotient will run into float precision issues and you actually should use a larger epsilon, for example. Also, there are suggestions that the symmetric difference quotient (f(x+eps) - f(x-eps))/(e*eps) often works a bit more reliably than the one-sided difference quotient this PR is using. So which should be used for a first "quick and dirty" numerical differentiation, if that's the way you want to go? That's why I was suggesting that since this approach can now be used by anyone with a single-line function definition `nd(func, x) = (func(x+epsilon)-func(x-epsilon))/(2*epsilon)` this would go better as a new item in the collection of examples (say "higher order functions and numerical differentiation" as a title for the entry), rather than as an officially blessed mathjs function, until such time as someone has made the time and effort to actually learn about current numeric differentiation techniques and implement something more sophisticated and robust -- but that there may not ever really be a need for it, if we just somewhat enhance mathjs's existing symbolic/automatic differentation. There really shouldn't be any functions defined in mathjs it can differentiate, and it's even plausible to extend that approach to plain javascript functions.

Anyhow, I think I've gone over this plenty now, so I will leave the disposition of this PR to you.

**Comment by josdejong** [comment]
Thanks for your inputs Glen, that's really helpful 👍. Here my considerations based on your comments:

1. We can already do symbolic differentiation in many cases, and that is highly preferable over numeric differentiation whenever available
2. The current implementation of `derivative` in this PR is indeed a one-liner, so anyone who needs it can easily implement it himself. So there is not much added value at this moment, whilst the current implementation does come with reliability issues.
3. In my head, the numerical differentiation would go hand-in-hand with numerical integration (which we do not yet have). As far as I know, integration is much harder to do symbolically, so I expect that there is more need for numerical integration. 
4. Let's postpone implementing `derivativeNum` until we do have a more sophisticated implementation, and implement it together with a corresponding `integrateNum`.

Are you OK with that @jjram20 ?

**Comment by gwhitney** [comment]
> 3\. As far as I know, integration is much harder to do symbolically, so I expect that there is more need for numerical integration.

Yes, I completely agree with that sentiment.

**Comment by josdejong** [comment]
I'll now close this PR. @jjram20 feel free to reopen/comment if you have more thoughts on this. And thanks for your efforts!

**Pull request closed without merging**